[ERP-520] Upgrade to Django 4

ScriptSmith opened the pull request.

ScriptSmith committed:
Cache python dependencies during docker build

ScriptSmith committed:
Update aloe

Replace aloe harvest command with GherkinTestRunner

https://docs.djangoproject.com/en/5.0/releases/3.2/#features-deprecated-in-3-2

> Using a boolean value in BaseCommand.requires_system_checks is deprecated. Use '__all__' instead of True, and [] (an empty list) instead of False.

ScriptSmith committed:
Update graphene-django 2.1.15 -> 3.1.5

ScriptSmith committed:
Specify format in usage of date_format

https://docs.djangoproject.com/en/4.2/releases/4.0/#time-zone-support
> In order to follow good practice, the default value of the USE_TZ setting will change from False to True, and time zone support will be enabled by default, in Django 5.0.

https://docs.djangoproject.com/en/4.2/ref/settings/#std-setting-USE_TZ
> In Django 5.0, the default value will change from False to True.

ScriptSmith committed:
Update djangorestframework 3.13.1 -> 3.14.0

https://github.com/encode/django-rest-framework/releases/tag/3.13.0

> Django 4.0 compatability

ScriptSmith committed:
Remove SECURE_BROWSER_XSS_FILTER setting

https://docs.djangoproject.com/en/4.0/releases/4.0/#securitymiddleware-no-longer-sets-the-x-xss-protection-header

> Most modern browsers don’t honor the X-XSS-Protection HTTP header. You can use Content-Security-Policy without allowing 'unsafe-inline' scripts instead.

ScriptSmith committed:
Update to Django 4.2

https://docs.djangoproject.com/en/5.0/releases/4.2/#dropped-support-for-postgresql-11

> Upstream support for PostgreSQL 11 ends in November 2023. Django 4.2 supports PostgreSQL 12 and higher.

ScriptSmith committed:
Remove unused countries tag

(templates.E003) 'countries' is used for multiple template tag modules: 'django_countries.templatetags.countries', 'rdrf.templatetags.countries'

ScriptSmith committed:
Update django-two-factor-auth 1.13.2 -> 1.15.5

https://github.com/jazzband/django-two-factor-auth/blob/master/CHANGELOG.md#1140

> Additionally, as the two_factor templatetags library was only containing phone-related filters, the library was renamed to phonenumber.

ppettitau reviewed (changes requested):
1. Date format in patient list for "Date of Birth" column differs 
    Local: 
    ![image](https://github.com/eresearchqut/trrf/assets/60163872/c7d55085-9eb3-4d52-adc1-e9ef5b307edc)
    Dev: 
    ![image](https://github.com/eresearchqut/trrf/assets/60163872/1a0dc825-0035-435f-8108-e089c74258ef)

ScriptSmith commented:
> 1. Date format in patient list for "Date of Birth" column differs
>        Local:
>        ![image](https://private-user-images.githubusercontent.com/60163872/307650184-c7d55085-9eb3-4d52-adc1-e9ef5b307edc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwMDkwNDQsIm5iZiI6MTcwOTAwODc0NCwicGF0aCI6Ii82MDE2Mzg3Mi8zMDc2NTAxODQtYzdkNTUwODUtOWViMy00ZDUyLWFkYzEtZTllZjViMzA3ZWRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI3VDA0MzkwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkZDdlMmIzN2Y4N2QxZjk0ZmE2MmE1Y2IxMGMwZDI4MGE2ZjM0YmI0YzA2NDcwNDdlMzE0NjMzMmQwNDFiOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZaIdwdIISKbG7IlARhrd_jVqZ_GfpPcbc1sJiQYApZY)
>        Dev:
>        ![image](https://private-user-images.githubusercontent.com/60163872/307650229-1a0dc825-0035-435f-8108-e089c74258ef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwMDkwNDQsIm5iZiI6MTcwOTAwODc0NCwicGF0aCI6Ii82MDE2Mzg3Mi8zMDc2NTAyMjktMWEwZGM4MjUtMDAzNS00MzVmLTgxMDgtZTA4OWM3NDI1OGVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI3VDA0MzkwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwYmUwNDNhOWRkOWViM2Q3OGM4M2NjZjU5ZmU5MTdiMTI2YjVjZDIxY2FjODY2MTcyYmY0MWFhZmM3NjFkNDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IzabRe7ePjt0dwXT5pnM9dVgd1Ov2mMZh-7njbNWUSQ)

:+1: I've updated the other usages of date_format except for 1 which is superficial and only visible to admins.

ScriptSmith committed:
Additional date_format format updates

ScriptSmith committed:
No-op migration

https://docs.djangoproject.com/en/4.0/releases/4.0/#migrations-autodetector-changes

> As a side-effect, running makemigrations might generate no-op AlterField operations for ManyToManyField and ForeignKey fields in some cases.
